app/controllers/users_controller.rb

```diff
@@ -32,11 +32,38 @@ def study_reports
     @study_reports = StudyReport.for_user(@user).page(params[:page]).order(id: :desc)
   end
 
+  def print_swipecard
+    swipecard = params[:swipecard]
+    printer = params[:printer]
+    if swipecard.strip.present?
+      print_swipecard_with_pmb(swipecard, printer)
+    else
+      flash[:error] = 'Cannot print empty swipecard'
+    end
+    redirect_to action: :show, id: @user.id
+  end
+
   private
 
   def find_user
     @user = User.find(params[:id])
   end
+  
+  def print_swipecard_with_pmb(swipecard, printer)
+    LabelPrinter::PmbClient.print({ 
```

Comment:
Most other controllers interact with `PrintJob#execute` which gives you your error handling, and centralises all our label classes in one place. This makes it much easier to review how printing it being used, and keeps the controllers nice and light.

Reply:
Part of me also wonders if this should be an entirely separate controller, but perhaps that's something we can put off until we rationalise the rest of our printing. I've got a branch in progress at the moment that is stripping out a load of the legacy batch stuff, which will put us in a much stronger position.